### ISSUE TITLE
Can recorder supports to record audio while connecting external display as well ?

### ISSUE THREAD

**Issue by dreampiggy** [m-rzb/screen-recorder-for-steamdeck#2]
I tested in SteamOS 3.3.3, insteall and launch in Game Mode (then keep in background), actually can produce a mp4 file inside `~/Videos`, but opened with no sound (using MPV player)

**Comment by dreampiggy** [comment]
More test environment:

I connect Steam Deck via USB-C dock to output to 1080P to my external display, and set it as primary display, does this cause issues ?

**Comment by dreampiggy** [comment]
Found one strange hardcode audio device in script source code here... https://github.com/m-rzb/Screen-recorder-for-steamdeck/blob/main/recapture-cli-wrapper.sh#L39

**Comment by dreampiggy** [comment]
aplay output:

```
(deck@steamdeck ~)$ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: Generic [HD-Audio Generic], device 3: HDMI 0 [HDMI 0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: Generic [HD-Audio Generic], device 7: HDMI 1 [HDMI 1]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: Generic [HD-Audio Generic], device 8: HDMI 2 [HDMI 2]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: Generic [HD-Audio Generic], device 9: HDMI 3 [HDMI 3]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: acp5x [acp5x], device 0: Playback/Capture nau8821-hifi-0 []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: acp5x [acp5x], device 1: CS35L41 Stereo Playback multicodec-1 []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

```
(deck@steamdeck ~)$ pw-cli list-objects | grep node.name
                node.name = "Dummy-Driver"
                node.name = "Freewheel-Driver"
                node.name = "echo-cancel-sink"
                node.name = "filter-chain-capture"
                node.name = "filter-chain-source"
                node.name = "Midi-Bridge"
                node.name = "alsa_output.pci-0000_04_00.5-platform-acp5x_mach.0.HiFi__hw_acp5x_1__sink"
                node.name = "alsa_output.pci-0000_04_00.1.hdmi-stereo-extra2"
                node.name = "alsa_input.pci-0000_04_00.5-platform-acp5x_mach.0.HiFi__hw_acp5x_0__source"
                node.name = "alsa_output.pci-0000_04_00.5-platform-acp5x_mach.0.HiFi__hw_acp5x_0__sink"
                node.name = "echo-cancel-capture"
                node.name = "echo-cancel-source"
                node.name = "echo-cancel-playback"
```

**Comment by dreampiggy** [comment]
I don't know more about linux audio device, can the script try to detech HDMI audio output ?

I run and find it actually use this device:

```
        id 63, type PipeWire:Interface:Node/3
                object.serial = "546"
                object.path = "alsa:pcm:0:hdmi:0,2:playback"
                factory.id = "18"
                client.id = "35"
                device.id = "53"
                priority.session = "900"
                priority.driver = "900"
                node.description = "HD-Audio Generic Digital Stereo (HDMI 3)"
                node.name = "alsa_output.pci-0000_04_00.1.hdmi-stereo-extra2"
                node.nick = "HDMI 2"
                media.class = "Audio/Sink"

```

![Screenshot_20230108_015739](https://user-images.githubusercontent.com/6919743/211164151-66f4ba2d-0a98-4fff-ac08-34482a7d235e.png)

**Comment by m-rzb** [comment]
I have not tested my Steam Deck with an externl display, to be honest. 



> aplay output:
> 
> ```
> (deck@steamdeck ~)$ aplay -l
> **** List of PLAYBACK Hardware Devices ****
> card 0: Generic [HD-Audio Generic], device 3: HDMI 0 [HDMI 0]
>   Subdevices: 1/1
>   Subdevice #0: subdevice #0
> card 0: Generic [HD-Audio Generic], device 7: HDMI 1 [HDMI 1]
>   Subdevices: 1/1
>   Subdevice #0: subdevice #0
> card 0: Generic [HD-Audio Generic], device 8: HDMI 2 [HDMI 2]
>   Subdevices: 1/1
>   Subdevice #0: subdevice #0
> card 0: Generic [HD-Audio Generic], device 9: HDMI 3 [HDMI 3]
>   Subdevices: 1/1
>   Subdevice #0: subdevice #0
> card 1: acp5x [acp5x], device 0: Playback/Capture nau8821-hifi-0 []
>   Subdevices: 1/1
>   Subdevice #0: subdevice #0
> card 1: acp5x [acp5x], device 1: CS35L41 Stereo Playback multicodec-1 []
>   Subdevices: 1/1
>   Subdevice #0: subdevice #0
> ```
> 
> ```
> (deck@steamdeck ~)$ pw-cli list-objects | grep node.name
>                 node.name = "Dummy-Driver"
>                 node.name = "Freewheel-Driver"
>                 node.name = "echo-cancel-sink"
>                 node.name = "filter-chain-capture"
>                 node.name = "filter-chain-source"
>                 node.name = "Midi-Bridge"
>                 node.name = "alsa_output.pci-0000_04_00.5-platform-acp5x_mach.0.HiFi__hw_acp5x_1__sink"
>                 node.name = "alsa_output.pci-0000_04_00.1.hdmi-stereo-extra2"
>                 node.name = "alsa_input.pci-0000_04_00.5-platform-acp5x_mach.0.HiFi__hw_acp5x_0__source"
>                 node.name = "alsa_output.pci-0000_04_00.5-platform-acp5x_mach.0.HiFi__hw_acp5x_0__sink"
>                 node.name = "echo-cancel-capture"
>                 node.name = "echo-cancel-source"
>                 node.name = "echo-cancel-playback"
> ```



> Found one strange hardcode audio device in script source code here... https://github.com/m-rzb/Screen-recorder-for-steamdeck/blob/main/recapture-cli-wrapper.sh#L39

It is not a strange audio device. Recapture go-binary has default audio device as "echo-cancel-sink.monitor". See the comments and OPTIONS for the recapture binary in the same script. Default audio source did not work on my Steam Deck. I looked on the internet as well. The default audio device on Steam Deck at the moment is "alsa_output.pci-0000_04_00.5-platform-acp5x_mach.0.HiFi__hw_acp5x_1__sink". You can see it in the output you posted above as well.

Try replacing the audio source with "alsa_output.pci-0000_04_00.1.hdmi-stereo-extra2" in $HOME/.local/recapture/recapture-cli-wrapper.sh on your Steam Deck, instead. I assume that is your external display.

**Comment by dreampiggy** [comment]
Ugly hack, but works...I don't know how to query "current KDE audio setting's first audio device", but I can enumerate using common order...

https://github.com/m-rzb/Screen-recorder-for-steamdeck/pull/3

**Comment by dreampiggy** [comment]
Here is my video and it has audio channel now.

Tested the order looks suitable, because HDMI display user will never use headphone, right :)

![Screenshot_20230108_022135](https://user-images.githubusercontent.com/6919743/211165101-e3d38a3a-d844-4945-b560-18cadb3a66d2.png)

**Comment by m-rzb** [comment]
> Here is my video and it has audio channel now.
> 
> Tested the order looks suitable, because HDMI display user will never use headphone, right :)
> 

Tested quickly your patch. With the Headphones audio source I had no audio when recording on my Steam Deck. After removing headphones, I had audio recorded again. Can you test it on your side as well, without connecting to external display?

**Comment by dreampiggy** [comment]
Strange, my headphone one (use idevice head phone) works...This match what I see in audio settings. I can upload you a video to approve.

Actually it should capture the `audio showed in desktop audio settings chosen one`. This just hack, and does not works for more cases, like bluetooth headphone.

---

![Screenshot_20230108_025936](https://user-images.githubusercontent.com/6919743/211166422-a2780370-90e0-4fdd-91e9-bef2b355d6ac.png)

record with sound: [2023-01-08_02-52-09.tar.gz](https://github.com/m-rzb/Screen-recorder-for-steamdeck/files/10367018/2023-01-08_02-52-09.tar.gz)

**Comment by m-rzb** [comment]
> Strange, my headphone one works...I can upload you a video.
> 
> Actually it should capture the `audio showed in desktop audio settings choose one`. This just hack, and does not works for more cases, like bluetooth headphone

Can you try to record the audio without the headset and external desplay pluged in/connected?

**Comment by dreampiggy** [comment]
When you click the audio icon, does the toast showing like switching to this device ? I always see this and I can hear a `noise` because of connecting.

**Comment by dreampiggy** [comment]
> Can you try to record the audio without the headset and external desplay pluged in/connected? 

Actually the first try was this case. I use Steam Deck as dock and connect to a HDMI display (no headphone here).

Then I test the Steam Deck without HDMI and only connect headphone.

Finally with Steam Deck with speaker (no any extra connects)

**Comment by dreampiggy** [comment]
Maybe this is just quick hack. Need someone who is expert at Linux audio to use some command lines or API to get exact what primary audio channel is. Maybe the recapture's author ?

**Comment by m-rzb** [comment]
I have no experience in coding. I created the script for my use case. If your hack works I will pull your request. But I will need to test it first..

**Comment by dreampiggy** [comment]
> Strange, my headphone one (use idevice head phone) works...This match what I see in audio settings. I can upload you a video to approve.
> 
> Actually it should capture the `audio showed in desktop audio settings chosen one`. This just hack, and does not works for more cases, like bluetooth headphone.
> 
> ![Screenshot_20230108_025936](https://user-images.githubusercontent.com/6919743/211166422-a2780370-90e0-4fdd-91e9-bef2b355d6ac.png)
> 
> record with sound: [2023-01-08_02-52-09.tar.gz](https://github.com/m-rzb/Screen-recorder-for-steamdeck/files/10367018/2023-01-08_02-52-09.tar.gz)

When you connect the headphone, can you click the audio icon in bottom bar, to see what shows here (any toasting ? ).  Doesn't it shows `headphone` instead ?

Or, is this game's issue (I guess, is Proton caused issue) ? Can the SteamOS GUI audio been record ?

---

And, SteamOS/KDE version ? I didn't upgrade to SteamOS 3.4.0 or KDE 25 because I heard some bugs effecy my use case and wait next release.

**Comment by m-rzb** [comment]
> Strange, my headphone one (use idevice head phone) works...This match what I see in audio settings. I can upload you a video to approve.

I believe you. 
I checked running the terminal in ssh, and the recording was via the headset, even though the headset was not plugged in. 
So then I also tested with the headphone plugged in: The recording had sound in SteamOS UI and in the games.
If I unplug the headphones, no audio in the recording again.

On my Steam Deck - SteamOs is 3.4.4 and KDE Plasma 5.26.1

Probably forcing the SteamOS using the "echo-cancel-sink" could be an option...

**Comment by m-rzb** [comment]
@dreampiggy can you try this https://raw.githubusercontent.com/m-rzb/Screen-recorder-for-steamdeck/testing-v2.1.0/recapture-cli-wrapper.sh
Please test it and let me know if it is working or not with your external display.

I used your idea, however, it has a different approach. To find the default sink, the patch is using `pactl get-default-sink`.
With my limited testing, with and without the headphones, it was picking the correct audio source each time and all recordings have audio. This approach will not change the audio source on the fly, meaning, if you unplug the headphone during recording it probably will be still recording the initial audio source...

**Comment by dreampiggy** [comment]
@m-rzb Tested that `pactl get-default-sink` works. Where did you find this :)

3 cases works
1. HDMI to external display
2. Headphone
3. Speaker

**Comment by dreampiggy** [comment]
> This approach will not change the audio source on the fly, meaning, if you unplug the headphone during recording it probably will be still recording the initial audio source...

It's OK in most cases, we can add this attention in the README.md
For fixing maybe this need some `event/notification` based solution to get changes for audio sink, but I don't know how

**Comment by m-rzb** [comment]
> @m-rzb Tested that `pactl get-default-sink` works. Where did you find this :)

`man pactl` :) The gstreamer implementation in Recapture is using pipewire with pulseaudio plugin. That's how I have identified the audio source when I was trying to figure out how to use the Recapture plugin two weeks ago. To be honest, I did not think about using it until receiving your pull request. Thank you for the idea!



> 3 cases works
> 
>     1. HDMI to external display
> 
>     2. Headphone
> 
>     3. Speaker

Great! Thank you for testing it. If it is working for you, I will apply the changes and push them later into `main`.

> It's OK in most cases, we can add this attention in the README.md

Will need to test, if it is actually affecting the recording or not. I haven't tested so cannot confirm. Mentioning that in the README.md shouldn't be a problem though.
